sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/ReadViaSDF.java

```diff
@@ -0,0 +1,861 @@
+      this.consumerConfig = transform.getConsumerConfig();
+      this.offsetConsumerConfig = transform.getOffsetConsumerConfig();
+      this.keyDeserializerProvider = transform.getKeyDeserializerProvider();
+      this.valueDeserializerProvider = transform.getValueDeserializerProvider();
+      this.consumerFactoryFn = transform.getConsumerFactoryFn();
+      this.extractOutputTimestampFn = transform.getExtractOutputTimestampFn();
+      this.createWatermarkEstimatorFn = transform.getCreateWatermarkEstimatorFn();
+      this.timestampPolicyFactory = transform.getTimestampPolicyFactory();
+    }
+
+    private final Map<String, Object> offsetConsumerConfig;
+
+    private final SerializableFunction<Map<String, Object>, Consumer<byte[], byte[]>>
+        consumerFactoryFn;
+    private final SerializableFunction<KafkaRecord<K, V>, Instant> extractOutputTimestampFn;
+    private final SerializableFunction<Instant, WatermarkEstimatorT> createWatermarkEstimatorFn;
+    private final TimestampPolicyFactory<K, V> timestampPolicyFactory;
+
+    // Variables that are initialized when bundle is started and closed when FinishBundle is called.
+    private transient ConsumerSpEL consumerSpEL = null;
+    private transient Deserializer<K> keyDeserializerInstance = null;
+    private transient Deserializer<V> valueDeserializerInstance = null;
+
+    private transient HashMap<TopicPartition, KafkaIOUtils.MovingAvg> avgRecordSize;
+    private transient HashMap<TopicPartition, KafkaIOUtils.MovingAvg> avgOffsetGap;
+
+    private static final Duration KAFKA_POLL_TIMEOUT = Duration.millis(1000);
+
+    @VisibleForTesting final DeserializerProvider keyDeserializerProvider;
+    @VisibleForTesting final DeserializerProvider valueDeserializerProvider;
+    @VisibleForTesting final Map<String, Object> consumerConfig;
+
+    /**
+     * A {@link GrowableOffsetRangeTracker.RangeEndEstimator} which uses a Kafka {@link Consumer} to
+     * fetch backlog.
+     */
+    private static class KafkaLatestOffsetEstimator
+        implements GrowableOffsetRangeTracker.RangeEndEstimator {
+
+      private final Consumer<byte[], byte[]> offsetConsumer;
+      private final TopicPartition topicPartition;
+      private final ConsumerSpEL consumerSpEL;
+      private final Supplier<Long> memorizedBacklog;
+
+      KafkaLatestOffsetEstimator(
+          Consumer<byte[], byte[]> offsetConsumer, TopicPartition topicPartition) {
+        this.offsetConsumer = offsetConsumer;
+        this.topicPartition = topicPartition;
+        this.consumerSpEL = new ConsumerSpEL();
+        this.consumerSpEL.evaluateAssign(
+            this.offsetConsumer, ImmutableList.of(this.topicPartition));
+        memorizedBacklog =
+            Suppliers.memoizeWithExpiration(
+                () -> {
+                  consumerSpEL.evaluateSeek2End(offsetConsumer, topicPartition);
+                  return offsetConsumer.position(topicPartition);
+                },
+                5,
```

Comment:
I forgot the default progress interval. Is it 5s or 5 mins?

Reply:
I believe the default is 30 seconds for Dataflow but I don't thing we should tune this to be Dataflow specific.

Reply:
I would be helpful to add a comment why this number was chosen.